docs/async-await.md

```diff
@@ -0,0 +1,168 @@
+# Proposal: Async-await support
+
+## Introduction
+
+With the introduction of [async/await][SE-0296] in Swift 5.5, it is now possible to write asynchronous code without the need for callbacks. 
+
+Language support for [`AsyncSequence`][SE-0298] also allows for writing functions that return values over time.
+
+We would like to explore how we could offer APIs that make use of these new language features to allow users to run HTTPRequest using these new idioms.
+
+This proposal describes what these APIs could look like and explores some of the potential usability concerns.
+
+## Proposed API additions
+
+### New `AsyncRequest` type
+
+The proposed new `AsyncRequest` shall be a simple swift structure.
+
+```swift
+struct AsyncRequest {
+  /// The requests url.
+  var url: String
+  
+  /// The request's HTTPMethod
+  var method: HTTPMethod
+  
+  /// The request's headers
+  var headers: HTTPHeaders
+
+  /// The request's body
+  var body: Body?
+    
+  init(url: String) {
+    self.url = url
+    self.method = .GET
+    self.headers = .init()
+    self.body = .none
+  }
+}
+```
+
+A notable change from the current [`HTTPRequest`][HTTPRequest] is that the url is not of type `URL`. This makes the creation of a request non throwing. Existing issues regarding current API: 
+
+- [HTTPClient.Request.url is a let constant][issue-395]
+- [refactor to make request non-throwing](https://github.com/swift-server/async-http-client/pull/56)
+- [improve request validation](https://github.com/swift-server/async-http-client/pull/67)
+
+The url validation will become part of the normal request validation that occurs when the request is scheduled on the `HTTPClient`. If the user supplies a request with an invalid url, the http client, will reject the request.
```

Comment:
If you want the URL validation to be part of request creation, you could create a wrapper which initializes a request from a URL object but creates the request using a String. That should give you a reasonable guarantee that HTTPClient will not reject the request due to it being an invalid URL.

It would mean parsing the URL twice, but that’s relatively cheap, and there are potentially ways to avoid it if we get shared Strings in the standard library at some point (they are already part of the ABI). If the shared String is owned by a URL object, the URL parser can assume it is already a valid URL and recover the parsed URL object from the owner without parsing the String again or allocating fresh storage.